### ISSUE TITLE
Crash when using from multiple threads

### ISSUE THREAD

**Issue by jw3126** [JuliaIO/HDF5.jl#835]
The following code crashes julia:
```julia
using HDF5
dir = tempdir()
@sync for i in 1:10
    data = randn(10)
    Threads.@spawn begin
        path = joinpath(dir, "$i.h5")
        h5open(path, "w") do file
            file["data"] = data
        end
    end
end
```
```
...
Bus error (core dumped)
```
Is there something that can be done to prevent the crash? Like putting a lock in `HDF5.jl`?

**Comment by simonbyrne** [comment]
By default, HDF5 is not thread-safe: there is an option to build it thread-safe, but this only serializes access anyway, and is incompatible with Parallel HDF5:
https://portal.hdfgroup.org/display/knowledge/Questions+about+thread-safety+and+concurrent+access

Adding locking to every function call could impose significant overhead, so probably should not be the default. With Julia, it isn't too hard to handle the locking yourself:
```julia
using HDF5
dir = tempdir()
lk = ReentrantLock()
@sync for i in 1:10
    data = randn(10)
    Threads.@spawn begin
        path = joinpath(dir, "$i.h5")
        lock(lk) do
            h5open(path, "w") do file
                file["data"] = data
            end
        end
    end
end

**Comment by jw3126** [comment]
Thanks! The problem in my case is, that there are two different libraries, that don't know about each other and both want to use HDF5. Currently, I impose the lock around functions like:
```julia
function some_library_function()
    while (...)
        perfectly_thread_safe_stuff(...)
        h5open(...)
    end
end
```
but this is wasteful.

**Comment by gszep** [comment]
should `HDF5_jll` build with `--enable-threadsafe` by default? I think https://github.com/PyTables/PyTables/issues/776 do this

**Comment by mkitti** [comment]
It would be good for us to consider adding thread safety features, but I would prefer to implement on the Julia side.

That said there should be a documented way to plugin the HDF5 C library with thread safety enabled there.

**Comment by simonbyrne** [comment]
What if we were to have a default lock which we use for finalizers (which would help with #990), and that could be acquired in other cases?

**Comment by mkitti** [comment]
For the lock to be effective, the user would also need to use the lock when using HDF5. That said, implementing a lock for finalizers is a good and perhaps necessary first step.

**Comment by gszep** [comment]
does concurrent read access work?
```julia
h5open(path, "r") do file
    Threads.@threads for id ∈ ids
        dataset = file["$id/dataset"][:]
        ############### concurrent lazy access and computations
        #
        #
    end
end
```

**Comment by mkitti** [comment]
> does concurrent read access work?

No, not through the HDF5 library via threads. You might be able to do it with memory mapping.

You can do concurrent reads with SWMR via multiple processes.

**Comment by gszep** [comment]
this MWE seems to work
```julia
function f()
    h5open("sample.h5", "r") do file
        for k ∈ keys(file)
            log10.(file["$k/counts"][:])
        end
    end
end

function g()
    h5open("sample.h5", "r") do file
        Threads.@threads for k ∈ keys(file)
            log10.(file["$k/counts"][:])
        end
    end
end

@time f() # 0.126000 seconds (2.75 k allocations: 112.420 MiB)
@time g() # 0.043185 seconds (2.84 k allocations: 112.429 MiB)
```

**Comment by mkitti** [comment]
It will work sometimes...

**Comment by simonbyrne** [comment]
@gszep What does your benchmark give if you use the `sb/lock` branch on #1021?

**Comment by simonbyrne** [comment]
Closed by #1021
